nixos/tests/grub.nix

```diff
@@ -0,0 +1,52 @@
+import ./make-test-python.nix ({ pkgs, ... }: {
+  name = "grub";
+
+  meta = with pkgs.stdenv.lib.maintainers; {
+    maintainers = [ rnhmjoj ];
+  };
+
+  machine = { lib, ... }: {
+    virtualisation.useBootLoader = true;
+
+    boot.loader.timeout = null;
+    boot.loader.grub = {
+      enable = true;
+      users.alice.password = "supersecret";
+
+      # OCR is not accurate enough
+      extraConfig = "serial; terminal_output serial";
+    };
+  };
+
+  testScript = ''
+    machine.start()
+
+    # wait for grub screen
+    machine.wait_for_console_text("GNU GRUB")
+
+    # select "All configurations" to trigger login request
+    machine.send_monitor_command("sendkey down")
+    machine.send_monitor_command("sendkey ret")
+
+    with subtest("Invalid credentials are rejected"):
+        machine.wait_for_console_text("Enter username:")
+        machine.send_chars("wronguser\n")
+        machine.wait_for_console_text("Enter password:")
+        machine.send_chars("wrongsecret\n")
+        machine.wait_for_console_text("error: access denied.")
+
+    # select "All configurations", again
+    machine.send_monitor_command("sendkey down")
+    machine.send_monitor_command("sendkey ret")
+
+    with subtest("Valid credentials are accepted"):
+        machine.wait_for_console_text("Enter username:")
+        machine.send_chars("alice\n")
+        machine.wait_for_console_text("Enter password:")
+        machine.send_chars("supersecret\n\n")
+        machine.wait_for_console_text("Linux version")
```

Comment:
wrapping these repeating lines into a function so that you can say

```python
select_all_configurations()
with subtest("Invalid credentials are rejected"):
        grub_login_as("wronguser", "wrongsecret")

select_all_configurations()
with subtest("Valid credentials are accepted"):
    grub_login_as("alice", "supersecret")
```

would make this test a nice example of how declarative integration test can be looking.

Reply:
Good idea, I will to it this way.